### PR TITLE
[REG-2018] Advanced Validations Guide

### PR DESCRIPTION
**Pull request by abeizer**
INSERT_PR_DETAILS_HERE

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

**Comment by RG-nAmKcAz** [comment]
I think the `Validation Suites` page should have a link to this new page.

Something like, looking for more advanced information about this topic ?  Reference our Advanced Validation Guide.

**Review by RG-nAmKcAz — APPROVED** [review]
This is fantastic.  I do have some suggestions to improve clarity and one concern about our SDK function, but overall .. gold star

File: docs/guides/writing-advanced-validations.mdx [hunk 1]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
 
-import Placeholder from '../partials/_placeholder.mdx';
+This guide provides in-depth examples for validating gameplay using [Validation Suites](../core-concepts/validation-suites).
 
-<Placeholder />
+:::note You will learn:
+✓ how to build validations for common scenarios such as navigating menus and exercising core gameplay mechanics<br/>
+✓ how to establish potential cause-and-effect relationships between actions and changes in the game state<br/>
+✓ how to build validations around non-deterministic behaviours
+:::
```
Comment (vontell) [inline docs/guides/writing-advanced-validations.mdx:13]: Love this idea of a primer, will add this to my own guides too.

File: docs/guides/writing-advanced-validations.mdx [hunk 2]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+
+**For this Gameplay Session**: we open the settings menu then increase the music volume and toggle graphics quality from "Medium" to "High".
+
+**In the Scenario Builder**: we'll validate that the settings were changed successfully.
+
+:::note Before Recording
+Set the [recording interval](../core-concepts/recording-gameplay#adjusting-the-recording-interval) to `1`. This is necessary for capturing changes to the settings menu.
+:::
+
+![Sample Gameplay Session](img/writing-advanced-validations/changing-in-game-settings/sample-gameplay-session.gif)
```
Comment (vontell) [inline docs/guides/writing-advanced-validations.mdx:39]: What tool did you end up using for the gifs? They look great!

File: docs/guides/writing-advanced-validations.mdx [hunk 3]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+        </figure>
+    </div>
+    <div className="col col--5">
+        <img src={StoreInitialVolumeSupplement} alt="Volume slider state"/>
+    </div>
+</div>
+
+Now, we need to validate that the volume increases.
+Add a "Wait for State" step and select the same `Handle` entity and its `position.x` field from the first two dropdowns.
+Select "greater than" (`>`) as your comparator.
```
Comment (vontell) [inline docs/guides/writing-advanced-validations.mdx:86]: Glad to see a comparison against a previous value in here!

File: docs/guides/writing-advanced-validations.mdx [hunk 4]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+and the second step waits for `shift` + `g` for "G".
+Then, we can validate that the `ProfilePopup`'s `InputText` component contains the text "RG".
+
+:::info Key Combinations
+Why can't we wait for `shift` + `r` + `g` in one "Wait for Key" step?
+
+This type of step validates that every key in the list is simultaneously held or not held.
+When users type, they tap keys, releasing one key before tapping the next -
+this is different from using key combinations during active gameplay
+(for example, `ctrl` + `shift` + `1` to cast a spell from your 20th hot-bar in your favorite MMO).
```
[the remark this answers is not in the document]
Comment (vontell) [reply docs/guides/writing-advanced-validations.mdx:164]: On my first read I thought it was a funny comment, was about to come here and say it'd be nice to show our gamer side in our docs. Open to Zack's comment though as well about where it should be used. But no change requested on my end at least.

**Review by vontell — APPROVED** [review]
Really no additional comments on top of Zack's that he already added. This guide is awesome! Very cool to see some very tricky and complicated tests in here. Especially seconding at least a few sentences to wrap it up.

**Review by addisonbgross — APPROVED** [review]
Aside from Zack's points, I think this guide is premium 🥇

At first glance I felt that it was too long, but I like how you've laid things out so that its a progressive example. Users can bail whenever they feel they've learned enough to get started 👍🏻 

(non-blocking at all) I think an internal guide to writing uhhh Guides like this would be an excellent IF thing. This is some top quality learning material

**Comment by batu** [comment]
Really great work! Only proposed change is maybe breaking it down into 3 sections instead of one mega section. I do remember there being a concern of too deeply nested pages, but I still think separating can work.

**Review by batu — APPROVED** [review]

File: docs/guides/writing-advanced-validations.mdx [hunk 5]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+
+**In the Scenario Builder**: we'll validate that destroying pots causes them to break into fragments,
+that contextual abilities change depending on whether the target is environmental or an NPC, and that
+enemies in an area can be damaged with AOE effects and die as a result of combat.
+
+![Sample Gameplay Session](./img/writing-advanced-validations/using-character-abilities/sample-gameplay-session.gif)
+
+The Scenario starts simply enough.
+Validate that the "BossRoom" dungeon scene is loaded and that our Archer character exists.
+Also assert that there are exactly five breakable pots in the game state (some of these are off-camera).
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:277]: These objects are unfortunately off-camera :-/

https://www.loom.com/share/19d23239c1424db583927ea7504a7c7b

File: docs/guides/writing-advanced-validations.mdx (hunk 2 above)
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:39]: [ScreenToGif](https://www.screentogif.com/)
It lets you extend and remove frames, add looping, etc. Was really nice for this!

File: docs/guides/writing-advanced-validations.mdx (hunk 5 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/guides/writing-advanced-validations.mdx:277]: Are you sure you're on a remotely recent sdk build then ?

I just checked and we definitely filter out objects that aren't within the visible screen bounds :/ They aren't considered for the bot_segments or data recording

[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:277]: Yes, I just pulled main and double-checked

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/guides/writing-advanced-validations.mdx:277]: ok, i recreated and will open a ticket to fix... that's going to messup your work in this PR, but the code is currently not working correctly :/

[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply docs/guides/writing-advanced-validations.mdx:277]: https://linear.app/regression-games/issue/REG-2130/off-screen-pots-being-included-in-state-of-bossroom

File: docs/guides/writing-advanced-validations.mdx [hunk 6]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
 
 # Writing Advanced Validations
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:5]: Added a link at the beginning of the "Creating Validation Scenarios" section in that page

File: docs/guides/writing-advanced-validations.mdx [hunk 7]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+
+Finish the Scenario by repeating this process for the text "High" instead of "Medium".
+
+### Final Scenario
+
+![Final Scenario](img/writing-advanced-validations/changing-in-game-settings/final-scenario.png)
+
+## Selecting a Character and Starting a Match
+
+This next example will validate user inputs in addition to interactions with UI components.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:126]: Added

File: docs/guides/writing-advanced-validations.mdx (hunk 4 above)
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx]: Rewrote this section to clarify. 

```
:::info Key Combinations
Why can't we wait for `shift` + `r` + `g` in one "Wait for Key" step?

This type of step validates that every key in the list is simultaneously held or not held.
When users type, they tap keys, releasing one key before tapping the next.
This input pattern differs from triggering gameplay mechanics like spells and abilities using key combinations.
For example, if you type "hello" into a chat-box then you'd expect to release "h" before pressing "e", and so on.
But you may need to hold an obscure combination of keys like `ctrl` + `shift` + `1` at the same time
to cast a spell from the 20th hot-bar in your favorite MMO.

When you use the "Wait for Key" step, consider what the player's intentions are when they produce those keyboard inputs.
:::
```

File: docs/guides/writing-advanced-validations.mdx [hunk 8]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+
+import TypeProfileNameStep from "./img/writing-advanced-validations/starting-a-match/type-profile-name-step.png"
+import TypeProfileNameSupplement from "./img/writing-advanced-validations/starting-a-match/type-profile-name-supplement.png"
+
+<div className="text--center margin-bottom--md">
+    <img src={TypeProfileNameStep} alt="Assert the user typed 'RG'" style={{width: "45rem"}}/>
+    <img src={TypeProfileNameSupplement} alt="'RG' is entered into text input" style={{width: "45rem"}}/>
+</div>
+
+Creating a new profile in Boss Room will automatically close the `ProfilePopup` dialog and activate that profile.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx:183]: This was in my final scenario but not these intermediate images 🤔 fixed

File: docs/guides/writing-advanced-validations.mdx [hunk 9]
```diff
@@ -4,6 +4,489 @@ sidebar_label: 'Writing Advanced Validations'
...
+but the fill level doesn't translate well to a meaningful health value.
+
+There may be many different ways to achieve a Scenario for your game.
+The "best" configuration of steps will depend heavily on your game state.
+
+:::
+
+Depending on the goals of your Scenario, you may want to repeat these steps for each cast of the AOE skill.
+You can repeat the last sequence of steps two more times to achieve this effect.
+Or, you may simply want to validate that the skill is mechanically sound and eventually results in the decimation of your enemies.
```
[the remark this answers is not in the document]
Comment (abeizer) [reply docs/guides/writing-advanced-validations.mdx]: Rewrote

```
For the sake of this example, we'll only write steps for the first AOE effect.
However, it would be good practice for this Scenario to repeat the same steps for each instance of the skill
to validate that it produces consistent results.
You can achieve this by repeating the previous sequence of steps two more times.

Finally, wrap up the Scenario by validating that all three Imps die at the same time and produce an `FX_ImpDefeat` animation.
```
